### PR TITLE
s4623205: Improved UNet

### PR DESCRIPTION
**Pull request by cloudstrife1117**
Dear Dr. Shakes and tutors,

Please review my pull request which contains 18 files in total, performing image segmentation on the skin lesions ISICs dataset with the Improved UNet. Below are the informations. Author, which includes who am I and my contact details. Objective, the target task to be completed. Files Descriptions, which contains 18 files in this pull request and descriptions of the files. Results, the final result of the task performance on the test set.

**Author:**
* **Name:** Jeng-Chung, Lien
* **Student ID:** 46232050
* **Email:** jengchung.lien@uqconnect.edu.au

**Objective:** Achieve a dice coefficient greater than 0.8 on ISICs dataset with Improved UNet performing segmentation.

**Files Descriptions:**
* **ExampleImage:** The folder that contains example images of the dataset, model architecture and the experiment results used in the report.
* **Models:** The folder that contains the model weights.
    * **Imporoved_Unet.h5:** The best Improved UNet model weights file trained in this project.
* **Modules:** The folder that contains modules.
    * **SegmentaionModel.py:** The module that contains the original Unet implementation in [[2]](https://arxiv.org/abs/1505.04597), Improved Unet implementation in [[1]](https://arxiv.org/abs/1802.10508v1). Functions to train the model with choices of using adam optimizers with decouple weight decay, learning rate decay or activate saving the best model and loading models.
    * **SegmentationMetrics.py:** The module that contains the evaluation metrics and loss functions used on segmentation tasks. Function of dice coefficient and dice loss.
    * **data_utils.py:** The module that contains functions to load and preprocess data.
    * **misc_utils.py:** The module that contains misc functions like print progress bar, get the maximum square image size that could fit in an image, and combine the image and segmentation mask function.
* **test_ImprovedUnet_ISICs.py:** The driver script that contains example on how to predict on the test set with loading the best Improved UNet model weights(**Used in this project to evaluate best model**).
* **train_ImprovedUnet_ISICs.py:** The driver script that contains example on how to train the Improved UNet model without saving the model weights and directly predict on the test set using the last epoch.
* **train_ImprovedUnet_ISICs_save.py:** The driver script that contains example on how to train the Improved UNet model with saving the model weights(**Used in this project to train and save best model**).

**Results:** A dice coefficient of **0.8894295** is achieved on the test set using the Improved Unet.

The README.md provides explanations of the task, the dataset, the model, techniques used on training the model and examples of how to use the model.
I haved spent many hours working on this project. I'm looking forward to your feedback and adjust. Thank you!

Kind Regards,
Jeng-Chung Lien

### CODE REVIEW

**Comment by wooboyeong** [comment]
General Comments
* Code OK for EASY
* Commit messages OK
* Documentation OK
* Pull request OK

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage 
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Please remove model files, these will clog up the repository. Your pull request will not be merged until they are removed. Feedback marks will be awarded when these are removed. -2
Request Description very good

**Comment by cloudstrife1117** [comment]
> ## TF/Torch Usage
> Adequate use and implementation
> 
> ## Good Practice (Design/Commenting)
> Good spacing and comments Header blocks
> 
> ## Algorithm
> Driver Script present Shows Usage & Demo & Visualisation & Data usage Module present No Data leakage Difficulty: Easy -10
> 
> ## Commit Log
> Meaningful commit messages Progressive commits used
> 
> ## Documentation
> Good Description and Comments Markdown used PDF submitted
> 
> ## Pull Request
> Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) Please remove model files, these will clog up the repository. Your pull request will not be merged until they are removed. Feedback marks will be awarded when these are removed. -2 Request Description very good

Dear Dr. Shakes,

As requested I have removed the model file in "Models/Imporoved_Unet.h5", which would become 17 files in total. (**Commit log here [link](https://github.com/shakes76/PatternFlow/pull/288/commits/efd9bb923c8112597d819e02a3f4a54348523da7)**)
I have also updated the file descriptions in README.md.

Please let me know if such changes is acceptable, and if other changes are needed.
Thank you!

**Comment by shakes76** [comment]
Feedback marks awarded.
